
# **<span style = "color: plum"> Analizando los datos </span>**
# <span style = "color: plum"> Generando Estadísticas </span>

## Cargando los datos


In [218]:
import pandas as pd

In [219]:
df = pd.read_csv("base_datos_ventas_tratada.csv",parse_dates=["dt_venta", "dt_entrega"], index_col= "dt_venta")

In [220]:
df.head()

,matricula_funcionario,nombre_funcionario,cargo,codigo_tienda,nombre_tienda,codigo_producto,descripcion_producto,categoria,precio_costo,valor_unitario,cantidad,dt_entrega,valor_final
dt_venta,,,,,,,,,,,,,
2020-08-25 10:04:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10013,Fogón,Electrodomésticos,210.9,404.1,3,2020-08-31 16:39:00,1212.3
2020-06-10 15:34:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10011,Licuadora,Cocina,59.9,142.4,2,2020-06-11 13:59:00,284.8
2019-03-21 15:58:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10013,Fogón,Electrodomésticos,210.9,404.1,2,2019-03-28 09:44:00,808.2
2019-09-16 14:46:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10004,Máquina de lavar ropa,Electrodomésticos,1099.0,1709.1,2,2019-09-20 09:22:00,3418.2
2019-11-25 16:28:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10005,Notebook,Electrónicos,2299.9,3199.0,2,2019-11-26 13:42:00,6398.0


In [221]:
df = df.astype({"matricula_funcionario": "str", "matricula_funcionario": "str", "codigo_producto":"str"})

In [222]:
df.info()


<class 'pandas.DataFrame'>
DatetimeIndex: 3000 entries, 2020-08-25 10:04:00 to 2019-03-18 15:40:00
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   matricula_funcionario  3000 non-null   str           
 1   nombre_funcionario     3000 non-null   str           
 2   cargo                  3000 non-null   str           
 3   codigo_tienda          3000 non-null   str           
 4   nombre_tienda          3000 non-null   str           
 5   codigo_producto        3000 non-null   str           
 6   descripcion_producto   3000 non-null   str           
 7   categoria              3000 non-null   str           
 8   precio_costo           3000 non-null   float64       
 9   valor_unitario         3000 non-null   float64       
 10  cantidad               3000 non-null   int64         
 11  dt_entrega             3000 non-null   datetime64[us]
 12  valor_final            3000 non-null 

In [223]:
df.describe()[["precio_costo", "cantidad","valor_unitario","valor_final"]]

,precio_costo,cantidad,valor_unitario,valor_final
count,3000.000000,3000.000000,3000.000000,3000.000000
mean,545.512300,2.453333,926.941353,2234.948870
min,27.900000,1.000000,55.900000,55.900000
25%,109.900000,1.000000,189.900000,427.200000
50%,210.900000,2.000000,404.100000,1346.900000
75%,799.000000,3.000000,1424.910000,3353.800000
max,2299.900000,4.000000,3199.000000,12796.000000
std,563.133526,1.107955,841.839695,2449.473528


In [224]:
df.describe(include=["int","float"]).round(2)

,precio_costo,valor_unitario,cantidad,valor_final
count,3000.00,3000.00,3000.00,3000.00
mean,545.51,926.94,2.45,2234.95
std,563.13,841.84,1.11,2449.47
min,27.90,55.90,1.00,55.90
25%,109.90,189.90,1.00,427.20
50%,210.90,404.10,2.00,1346.90
75%,799.00,1424.91,3.00,3353.80
max,2299.90,3199.00,4.00,12796.00


## <span style = "color: plum"> [df|series].count() </span>

<span style = "color: plum"> Método 
</span> que cuenta la cantidadde valores no nulos de una variable

In [225]:
df.count()

matricula_funcionario    3000
nombre_funcionario       3000
cargo                    3000
codigo_tienda            3000
nombre_tienda            3000
codigo_producto          3000
descripcion_producto     3000
categoria                3000
precio_costo             3000
valor_unitario           3000
cantidad                 3000
dt_entrega               3000
valor_final              3000
dtype: int64

In [226]:
df["codigo_producto"].count()

np.int64(3000)

## <span style = "color: plum"> [df|series].sum() </span>

<span style = "color: plum"> Método 
</span> que suma los valores de una variable.

In [227]:
df.sum

<bound method DataFrame.sum of                     matricula_funcionario nombre_funcionario     cargo  \
dt_venta                                                                 
2020-08-25 10:04:00             123456999      Felipe Mariño  Vendedor   
2020-06-10 15:34:00             123456999      Felipe Mariño  Vendedor   
2019-03-21 15:58:00             123456999      Felipe Mariño  Vendedor   
2019-09-16 14:46:00             123456999      Felipe Mariño  Vendedor   
2019-11-25 16:28:00             123456999      Felipe Mariño  Vendedor   
...                                   ...                ...       ...   
2019-04-17 15:36:00             123457113        Victor Hugo  Vendedor   
2020-01-21 15:46:00             123457113        Victor Hugo  Vendedor   
2020-10-22 17:06:00             123457113        Victor Hugo  Vendedor   
2019-02-18 16:04:00             123457113        Victor Hugo  Vendedor   
2019-03-18 15:40:00             123457113        Victor Hugo  Vendedor   

      

In [228]:
columnas= ["precio_costo","valor_unitario","valor_final"]
df[columnas].sum()
#Me devuelve la suma total de cada una de las columnas, es decir, el total de precio costo, cantidad, valor unitario y valor final.

precio_costo      1636536.90
valor_unitario    2780824.06
valor_final       6704846.61
dtype: float64

In [229]:
#Si tengo 20 columnas de tipo numérico y quiero obtener la suma total de cada una de ellas
df.sum(numeric_only=True)

precio_costo      1636536.90
valor_unitario    2780824.06
cantidad             7360.00
valor_final       6704846.61
dtype: float64

In [230]:
#Total de facturacion de las ventas
df["valor_final"].sum()

np.float64(6704846.609999999)

In [231]:
df.head()

,matricula_funcionario,nombre_funcionario,cargo,codigo_tienda,nombre_tienda,codigo_producto,descripcion_producto,categoria,precio_costo,valor_unitario,cantidad,dt_entrega,valor_final
dt_venta,,,,,,,,,,,,,
2020-08-25 10:04:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10013,Fogón,Electrodomésticos,210.9,404.1,3,2020-08-31 16:39:00,1212.3
2020-06-10 15:34:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10011,Licuadora,Cocina,59.9,142.4,2,2020-06-11 13:59:00,284.8
2019-03-21 15:58:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10013,Fogón,Electrodomésticos,210.9,404.1,2,2019-03-28 09:44:00,808.2
2019-09-16 14:46:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10004,Máquina de lavar ropa,Electrodomésticos,1099.0,1709.1,2,2019-09-20 09:22:00,3418.2
2019-11-25 16:28:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10005,Notebook,Electrónicos,2299.9,3199.0,2,2019-11-26 13:42:00,6398.0


In [232]:
#Indicandole a pandas que me sume por el eje vertical o Horizontal

df.sum(numeric_only= True, axis=0) #Suma por el eje vertical, es decir, me devuelve la suma total de cada una de las columnas


precio_costo      1636536.90
valor_unitario    2780824.06
cantidad             7360.00
valor_final       6704846.61
dtype: float64

In [233]:
df.sum(numeric_only= True, axis=1) #Suma por el eje horizontal, es decir, me devuelve la suma total de cada una de las filas

dt_venta
2020-08-25 10:04:00     1830.3
2020-06-10 15:34:00      489.1
2019-03-21 15:58:00     1425.2
2019-09-16 14:46:00     6228.3
2019-11-25 16:28:00    11898.9
                        ...   
2019-04-17 15:36:00    10691.4
2020-01-21 15:46:00      861.6
2020-10-22 17:06:00      681.7
2019-02-18 16:04:00     1020.1
2019-03-18 15:40:00     7938.4
Length: 3000, dtype: float64

## <span style = "color: plum"> Estadística Descriptiva  </span>

## <span style = "color: plum">Media  Mean() </span>

In [234]:

df.mean(numeric_only= True, axis=0) #Me devuelve el promedio de cada una de las columnas

precio_costo       545.512300
valor_unitario     926.941353
cantidad             2.453333
valor_final       2234.948870
dtype: float64

In [235]:
df[["cantidad", "valor_final"]].mean()

cantidad          2.453333
valor_final    2234.948870
dtype: float64

In [236]:
df.mean(numeric_only = True)

precio_costo       545.512300
valor_unitario     926.941353
cantidad             2.453333
valor_final       2234.948870
dtype: float64

## <span style = "color: plum">Mediana  Median() </span>

In [237]:
df.median(numeric_only=True)

precio_costo       210.9
valor_unitario     404.1
cantidad             2.0
valor_final       1346.9
dtype: float64

## <span style = "color: plum">Moda  Mode() </span>

In [238]:
df.mode(numeric_only=True)

,precio_costo,valor_unitario,cantidad,valor_final
0,88.9,189.9,1,379.8


In [239]:
df["precio_costo"].mode()

0    88.9
Name: precio_costo, dtype: float64

In [240]:
df.columns

Index(['matricula_funcionario', 'nombre_funcionario', 'cargo', 'codigo_tienda',
       'nombre_tienda', 'codigo_producto', 'descripcion_producto', 'categoria',
       'precio_costo', 'valor_unitario', 'cantidad', 'dt_entrega',
       'valor_final'],
      dtype='str')

In [241]:
df["descripcion_producto"].mode()

0    Aspiradora de piso
Name: descripcion_producto, dtype: str

In [242]:
df["valor_final"].quantile([0.25, 0.5, 0.75]) #Me devuelve el valor del percentil 50, es decir, la mediana de la columna valor final

0.25     427.2
0.50    1346.9
0.75    3353.8
Name: valor_final, dtype: float64

## <span style = "color: plum">Rango intercuartilico</span>

In [243]:
df["valor_final"].quantile(0.75) - df["valor_final"].quantile(0.25)

np.float64(2926.6)

In [244]:
df[["cantidad", "valor_final"]].mean() #Me devuelve el promedio de cantidad y valor final

cantidad          2.453333
valor_final    2234.948870
dtype: float64

# <span style = "color: plum">Medidas de Dispersion </span>

## <span style = "color: plum">Varianza .var()</span>

In [245]:
df[["cantidad", "valor_final"]].var()

cantidad       1.227565e+00
valor_final    5.999921e+06
dtype: float64

In [246]:
df.std(numeric_only = True)

precio_costo       563.133526
valor_unitario     841.839695
cantidad             1.107955
valor_final       2449.473528
dtype: float64

## <span style = "color: plum"> Desviación Estandar .std()</span>

In [247]:
df[["cantidad", "valor_final"]].mean() #Me devuelve el promedio de cantidad y valor final

cantidad          2.453333
valor_final    2234.948870
dtype: float64

In [248]:
df[["cantidad", "valor_final"]].std()

cantidad          1.107955
valor_final    2449.473528
dtype: float64

## <span style = "color: plum"> Minimo .min()</span>

In [249]:
df[["cantidad", "valor_final"]].min()

cantidad        1.0
valor_final    55.9
dtype: float64

## <span style = "color: plum"> Maximo .max()</span>

In [250]:
df[["cantidad", "valor_final"]].max()

cantidad           4.0
valor_final    12796.0
dtype: float64

## <span style = "color: plum"> Rango</span>

In [251]:
df["precio_costo"].max() - df["precio_costo"].min()

np.float64(2272.0)

Significa que 2272 dolares , se puede evidenciar la disperdios  

## <span style = "color: plum">Correlacion y Covarianza </span>

Representan la relacion  o correlación entre variables

## <span style = "color: plum"> Covarianza .cov()</span>

In [252]:
# Extraer solo las columnas numéricas (enteros y decimales)
df_numerico = df.select_dtypes(include=['number'])


In [253]:
print(df.select_dtypes(include=['number']).columns.tolist())

['precio_costo', 'valor_unitario', 'cantidad', 'valor_final']


In [254]:
df_numerico.cov().round(2)

,precio_costo,valor_unitario,cantidad,valor_final
precio_costo,317119.37,469174.60,-24.23,1120710.35
valor_unitario,469174.60,708694.07,-39.16,1693575.30
cantidad,-24.23,-39.16,1.23,1067.18
valor_final,1120710.35,1693575.30,1067.18,5999920.56


In [255]:
df[["cantidad", "valor_final"]].cov()

,cantidad,valor_final
cantidad,1.227565,1.067182e+03
valor_final,1067.181589,5.999921e+06


## <span style="color: plum">¿Qué es la Covarianza? 📊</span>

La **Covarianza** es una medida estadística que indica la dirección de la relación lineal entre dos variables aleatorias. En términos simples, nos dice si dos variables "se mueven juntas" o en direcciones opuestas.

### 🧠 Interpretación de los Resultados

Para analizar nuestra matriz, debemos observar dos factores clave:

1. **El Signo (Dirección):**
    * **Positivo (+):** Las variables tienen una relación directa. Si una aumenta, la otra también (ej. *Precio Costo* y *Valor Final*).
    * **Negativo (-):** Las variables tienen una relación inversa. Si una aumenta, la otra disminuye (ej. *Precio* y *Cantidad*).
    * **Cero (0):** No existe una relación lineal aparente entre las variables.

2. **La Magnitud (Varianza):**
    * Los valores en la **diagonal principal** representan la **Varianza** de cada variable (qué tanto se alejan los datos de su propio promedio).



---

### 🔍 Análisis de mis Datos (Hallazgos Relevantes)

| Relación | Tipo de Covarianza | Significado en el Negocio |
| :--- | :--- | :--- |
| **Precios vs. Valor Final** | **Positiva Alta** | El modelo de precios es consistente; el aumento en costos se refleja directamente en el ingreso final. |
| **Cantidad vs. Valor Unitario** | **Negativa** | Existe una ligera tendencia donde los productos de mayor valor se venden en menores cantidades (Elasticidad). |
| **Diagonal Principal** | **Varianza** | Nos indica qué tan dispersos son los valores de cada columna por sí sola. |

> **Nota importante:** La covarianza depende de las unidades de medida (pesos vs. unidades), por lo que números muy grandes no siempre significan una relación más fuerte que los números pequeños.

## <span style = "color: plum"> Coheficinete de correlacion  .corr(method)</span>

In [256]:
df[["cantidad", "valor_final"]].corr()

,cantidad,valor_final
cantidad,1.000000,0.393227
valor_final,0.393227,1.000000


In [257]:
df_numerico.corr().round(2)

,precio_costo,valor_unitario,cantidad,valor_final
precio_costo,1.00,0.99,-0.04,0.81
valor_unitario,0.99,1.00,-0.04,0.82
cantidad,-0.04,-0.04,1.00,0.39
valor_final,0.81,0.82,0.39,1.00


## <span style="color: plum">¿Qué es la Correlación? 📈</span>

La **Correlación (de Pearson)** es una versión normalizada de la covarianza. Nos indica no solo la dirección de la relación, sino también la **fuerza** de la misma, eliminando el problema de las unidades de medida (pesos vs. unidades).

### 🧠 Escala de Interpretación

A diferencia de la covarianza, aquí los números siempre están entre **-1 y 1**:

* **1.0:** Correlación positiva perfecta (suben exactamente igual).
* **0.8 a 0.99:** Correlación muy fuerte.
* **0.0:** No hay relación lineal.
* **-1.0:** Correlación negativa perfecta (una sube y la otra baja proporcionalmente).


---

### 🔍 Análisis de mis Resultados (Correlación)

| Relación | Coeficiente | Interpretación Técnica |
| :--- | :--- | :--- |
| **Costo vs. Valor Unitario** | **0.99** | **Relación Casi Perfecta.** El valor unitario depende casi totalmente del costo de adquisición. |
| **Costo/Unitario vs. Valor Final** | **0.81 / 0.82** | **Relación Muy Fuerte.** El precio que paga el cliente está fuertemente ligado a los costos. |
| **Cantidad vs. Precios** | **-0.04** | **Relación Nula/Inexistente.** El precio del producto no parece afectar cuántas unidades se compran (comportamiento inelástico). |
| **Cantidad vs. Valor Final** | **0.39** | **Relación Débil.** El total de la venta sube un poco cuando hay más cantidad, pero los precios influyen más. |

### 💡 Conclusión Clave
Mientras que la covarianza nos daba números de millones que eran difíciles de comparar, la **correlación** nos confirma que la relación más sólida en tu dataset es entre el **costo y el valor unitario (0.99)**, y que la **cantidad** es la variable más independiente del grupo.

# <span style = "color: plum"> Filtrando los Datos</span>

In [258]:
df.head()

,matricula_funcionario,nombre_funcionario,cargo,codigo_tienda,nombre_tienda,codigo_producto,descripcion_producto,categoria,precio_costo,valor_unitario,cantidad,dt_entrega,valor_final
dt_venta,,,,,,,,,,,,,
2020-08-25 10:04:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10013,Fogón,Electrodomésticos,210.9,404.1,3,2020-08-31 16:39:00,1212.3
2020-06-10 15:34:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10011,Licuadora,Cocina,59.9,142.4,2,2020-06-11 13:59:00,284.8
2019-03-21 15:58:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10013,Fogón,Electrodomésticos,210.9,404.1,2,2019-03-28 09:44:00,808.2
2019-09-16 14:46:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10004,Máquina de lavar ropa,Electrodomésticos,1099.0,1709.1,2,2019-09-20 09:22:00,3418.2
2019-11-25 16:28:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10005,Notebook,Electrónicos,2299.9,3199.0,2,2019-11-26 13:42:00,6398.0


In [259]:
df.loc["2020-08-25 10:04:00"]

matricula_funcionario              123456999
nombre_funcionario             Felipe Mariño
cargo                               Vendedor
codigo_tienda                          CL003
nombre_tienda                      Filial SP
codigo_producto                        10013
descripcion_producto                   Fogón
categoria                  Electrodomésticos
precio_costo                           210.9
valor_unitario                         404.1
cantidad                                   3
dt_entrega               2020-08-31 16:39:00
valor_final                           1212.3
Name: 2020-08-25 10:04:00, dtype: object

In [260]:
df.describe().round(2)

,precio_costo,valor_unitario,cantidad,dt_entrega,valor_final
count,3000.00,3000.00,3000.00,3000,3000.00
mean,545.51,926.94,2.45,2019-12-28 13:30:03.600000,2234.95
min,27.90,55.90,1.00,2019-01-01 14:07:00,55.90
25%,109.90,189.90,1.00,2019-06-25 05:15:15,427.20
50%,210.90,404.10,2.00,2019-12-21 10:46:00,1346.90
75%,799.00,1424.91,3.00,2020-06-30 15:31:45,3353.80
max,2299.90,3199.00,4.00,2020-12-31 17:37:00,12796.00
std,563.13,841.84,1.11,NaN,2449.47


## <span style="color: plum">Operadores Relacionales ⚖️</span>

Los operadores relacionales son símbolos que se utilizan para comparar dos valores. El resultado de estas operaciones siempre es un valor booleano: `True` (Verdadero) o `False` (Falso).

### 📋 Lista de Operadores

* **Mayor que** `>`
* **Mayor o igual que** `>=`
* **Menor que** `<`
* **Menor o igual que** `<=`
* **Igual a (comparación)** `==`
* **Diferente de** `!=`


---

### 💡 Notas importantes para Python:
1. **Asignación vs. Comparación:** No confundas el `=` (que sirve para guardar un valor en una variable) con el `==` (que sirve para preguntar si dos cosas son iguales).
2. **Uso en DataFrames:** Estos operadores son fundamentales cuando queremos filtrar datos. 
   * *Ejemplo:* `df[df['precio_costo'] > 5000]` nos devolvería solo los productos caros.

In [261]:
df[df["valor_final"]>5000]

,matricula_funcionario,nombre_funcionario,cargo,codigo_tienda,nombre_tienda,codigo_producto,descripcion_producto,categoria,precio_costo,valor_unitario,cantidad,dt_entrega,valor_final
dt_venta,,,,,,,,,,,,,
2019-11-25 16:28:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10005,Notebook,Electrónicos,2299.9,3199.0,2,2019-11-26 13:42:00,6398.0
2019-12-14 13:44:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10005,Notebook,Electrónicos,2299.9,3199.0,2,2019-12-15 09:06:00,6398.0
2020-05-31 17:46:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10005,Notebook,Electrónicos,2299.9,3199.0,2,2020-06-01 15:56:00,6398.0
2019-08-10 16:08:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10004,Máquina de lavar ropa,Electrodomésticos,1099.0,1709.1,3,2019-08-13 14:12:00,5127.3
2020-02-21 14:19:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10001,Smartphone,Electrónicos,899.9,1676.9,3,2020-02-22 10:54:00,5030.7
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2019-09-11 16:53:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10005,Notebook,Electrónicos,2299.9,3199.0,2,2019-09-12 11:50:00,6398.0
2019-02-27 15:01:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10005,Notebook,Electrónicos,2299.9,3199.0,4,2019-02-28 14:23:00,12796.0
2020-08-13 13:06:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10001,Smartphone,Electrónicos,899.9,1676.9,4,2020-08-14 12:20:00,6707.6


In [262]:
filtro = df[df["valor_final"]>5000]
df_filtrado= filtro

## <span style="color: plum">Operadores Lógicos 🧠</span>

Los operadores lógicos nos permiten combinar múltiples comparaciones para crear filtros más complejos. Son la base para tomar decisiones en el código.

### 📋 Tabla de Operadores Lógicos

| Operador | Nombre | Descripción | Ejemplo |
| :--- | :--- | :--- | :--- |
| **and** (&) | **Y lógico** | Devuelve `True` si **ambos** operadores son `True`. | `True and False` -> **False** |
| **or** (\|) | **O lógico** | Devuelve `True` si **al menos uno** de los operadores es `True`. | `True or False` -> **True** |
| **not** | **Negación** | Devuelve lo **opuesto** del valor booleano del operando. | `not True` -> **False** |


---

### ⚠️ Importante para Pandas:
Cuando trabajas con DataFrames, Python requiere que uses los símbolos específicos en lugar de las palabras para que el filtro funcione correctamente en cada fila:
* Usa **`&`** en lugar de `and`.
* Usa **`|`** en lugar de `or`.
* Usa **`~`** en lugar de `not`.

**Ejemplo de filtro combinado:**
`df[(df['cantidad'] > 10) & (df['precio_costo'] < 5000)]`
*(Esto filtraría productos que son baratos Y de los cuales hay mucha cantidad).*

In [263]:
filtro_valor_final =df["valor_final"]>100

In [264]:
df["descripcion_producto"].unique()

<StringArray>
[                  'Fogón',               'Licuadora',
   'Máquina de lavar ropa',                'Notebook',
                 'Colchón',      'Freidora eléctrica',
               'Televisor',                    'Sofá',
 'Procesador de alimentos',                 'Plancha',
             'Guarda Ropa',                  'Tablet',
      'Aspiradora de piso',                'Heladera',
              'Smartphone',              'Microondas',
                'Cafetera',                'Batidora',
               'Impresora',                'Cama Box']
Length: 20, dtype: str

In [267]:
filtro_descripcion_producto = df["descripcion_producto"]== "Plancha"

In [269]:
df_valor_descProducto =df[filtro_valor_final & filtro_descripcion_producto]

In [271]:
print(df_valor_descProducto.shape[0])
df_valor_descProducto


123


,matricula_funcionario,nombre_funcionario,cargo,codigo_tienda,nombre_tienda,codigo_producto,descripcion_producto,categoria,precio_costo,valor_unitario,cantidad,dt_entrega,valor_final
dt_venta,,,,,,,,,,,,,
2019-06-21 10:21:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10009,Plancha,Casa,27.9,55.9,2,2019-06-21 10:31:00,111.8
2020-11-04 17:01:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10009,Plancha,Casa,27.9,55.9,2,2020-11-05 11:49:00,111.8
2019-07-20 16:04:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10009,Plancha,Casa,27.9,55.9,3,2019-07-21 11:57:00,167.7
2019-11-24 16:00:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10009,Plancha,Casa,27.9,55.9,3,2019-11-25 15:33:00,167.7
2019-01-02 13:55:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10009,Plancha,Casa,27.9,55.9,4,2019-01-02 14:58:00,223.6
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2019-11-19 14:08:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10009,Plancha,Casa,27.9,55.9,3,2019-11-20 12:35:00,167.7
2019-10-03 09:01:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10009,Plancha,Casa,27.9,55.9,4,2019-10-03 13:28:00,223.6
2019-03-04 14:21:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10009,Plancha,Casa,27.9,55.9,2,2019-03-04 14:44:00,111.8


In [272]:
df_valor_descProducto.to_csv("df_valor_descProducto.csv", index=False)

In [273]:
df[(df["descripcion_producto"]=="Microondas")  | (df["descripcion_producto"]=="Plancha")]

,matricula_funcionario,nombre_funcionario,cargo,codigo_tienda,nombre_tienda,codigo_producto,descripcion_producto,categoria,precio_costo,valor_unitario,cantidad,dt_entrega,valor_final
dt_venta,,,,,,,,,,,,,
2019-06-21 10:21:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10009,Plancha,Casa,27.9,55.9,2,2019-06-21 10:31:00,111.8
2019-03-30 11:14:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10015,Microondas,Electrodomésticos,189.0,364.5,3,2019-03-31 10:21:00,1093.5
2020-08-31 12:54:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10015,Microondas,Electrodomésticos,189.0,364.5,2,2020-09-01 09:20:00,729.0
2020-11-04 17:01:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10009,Plancha,Casa,27.9,55.9,2,2020-11-05 11:49:00,111.8
2019-09-03 12:17:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10015,Microondas,Electrodomésticos,189.0,364.5,3,2019-09-04 10:13:00,1093.5
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2020-05-06 16:18:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10009,Plancha,Casa,27.9,55.9,1,2020-05-07 13:47:00,55.9
2020-03-11 11:08:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10015,Microondas,Electrodomésticos,189.0,364.5,2,2020-03-11 15:14:00,729.0
2019-10-05 10:37:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10015,Microondas,Electrodomésticos,189.0,364.5,4,2019-10-05 12:52:00,1458.0


In [274]:
df_valor_descProducto.describe(include=["int","float"]).round(2)

,precio_costo,valor_unitario,cantidad,valor_final
count,123.0,123.0,123.00,123.00
mean,27.9,55.9,3.01,168.15
std,0.0,0.0,0.78,43.83
min,27.9,55.9,2.00,111.80
25%,27.9,55.9,2.00,111.80
50%,27.9,55.9,3.00,167.70
75%,27.9,55.9,4.00,223.60
max,27.9,55.9,4.00,223.60


## <span style="color: plum">Operadores Lógicos con `loc[ ]` 🔍</span>

Los operadores lógicos son fundamentales cuando utilizamos `.loc[]` para filtrar filas bajo múltiples condiciones. Nos permiten refinar nuestras búsquedas en el DataFrame de manera precisa.

### 📋 Tipos de Operadores para Filtrado

| Operador | Nombre | Descripción Técnica | Símbolo en Pandas |
| :--- | :--- | :--- | :--- |
| **and** | **Y lógico** | Devuelve `True` solo si **todas** las condiciones se cumplen. | `&` |
| **or** | **O lógico** | Devuelve `True` si **al menos una** de las condiciones se cumple. | `|` |
| **not** | **Negación** | Invierte el resultado (lo que era `True` pasa a ser `False`). | `~` |

### 💡 Ejemplo de Aplicación con `.loc`
```python
# Filtrar productos con costo > 1000 Y cantidad < 5
df_filtrado = df.loc[(df['precio_costo'] > 1000) & (df['cantidad'] < 5)]

In [275]:
df.head()   

,matricula_funcionario,nombre_funcionario,cargo,codigo_tienda,nombre_tienda,codigo_producto,descripcion_producto,categoria,precio_costo,valor_unitario,cantidad,dt_entrega,valor_final
dt_venta,,,,,,,,,,,,,
2020-08-25 10:04:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10013,Fogón,Electrodomésticos,210.9,404.1,3,2020-08-31 16:39:00,1212.3
2020-06-10 15:34:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10011,Licuadora,Cocina,59.9,142.4,2,2020-06-11 13:59:00,284.8
2019-03-21 15:58:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10013,Fogón,Electrodomésticos,210.9,404.1,2,2019-03-28 09:44:00,808.2
2019-09-16 14:46:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10004,Máquina de lavar ropa,Electrodomésticos,1099.0,1709.1,2,2019-09-20 09:22:00,3418.2
2019-11-25 16:28:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10005,Notebook,Electrónicos,2299.9,3199.0,2,2019-11-26 13:42:00,6398.0


In [277]:
#Primero filas y luego columnas
df.loc["2020-08-25 10:04:00", "nombre_funcionario"]

'Felipe Mariño'

In [279]:
df["codigo_producto"].unique()

<StringArray>
['10013', '10011', '10004', '10005', '10018', '10010', '10003', '10019',
 '10012', '10009', '10016', '10002', '10006', '10014', '10001', '10015',
 '10008', '10007', '10020', '10017']
Length: 20, dtype: str

In [283]:
# Los dos puntos indican que quiero todas las columnas, pero solo las filas donde el codigo_producto sea igual a 10016
df.loc[df["codigo_producto"]=="10016",:]

,matricula_funcionario,nombre_funcionario,cargo,codigo_tienda,nombre_tienda,codigo_producto,descripcion_producto,categoria,precio_costo,valor_unitario,cantidad,dt_entrega,valor_final
dt_venta,,,,,,,,,,,,,
2019-03-25 13:01:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10016,Guarda Ropa,Casa,489.9,999.0,2,2019-03-31 13:46:00,1998.0
2019-05-31 16:16:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10016,Guarda Ropa,Casa,489.9,999.0,2,2019-06-04 15:46:00,1998.0
2019-04-23 14:08:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10016,Guarda Ropa,Casa,489.9,999.0,3,2019-04-27 14:47:00,2997.0
2020-10-02 13:56:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10016,Guarda Ropa,Casa,489.9,999.0,2,2020-10-05 16:33:00,1998.0
2019-09-09 11:08:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10016,Guarda Ropa,Casa,489.9,999.0,3,2019-09-16 11:31:00,2997.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2019-01-31 09:39:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10016,Guarda Ropa,Casa,489.9,999.0,2,2019-02-03 12:25:00,1998.0
2019-08-19 14:25:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10016,Guarda Ropa,Casa,489.9,999.0,2,2019-08-21 12:37:00,1998.0
2020-05-04 17:54:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10016,Guarda Ropa,Casa,489.9,999.0,4,2020-05-10 12:48:00,3996.0


In [ ]:
df.loc[df["codigo_producto"]=="10016", "precio_costo"]

dt_venta
2019-03-25 13:01:00    489.9
2019-05-31 16:16:00    489.9
2019-04-23 14:08:00    489.9
2020-10-02 13:56:00    489.9
2019-09-09 11:08:00    489.9
                       ...  
2019-01-31 09:39:00    489.9
2019-08-19 14:25:00    489.9
2020-05-04 17:54:00    489.9
2019-11-08 09:59:00    489.9
2020-04-08 16:03:00    489.9
Name: precio_costo, Length: 149, dtype: float64

In [288]:
df.loc[df["codigo_producto"]=="10016", "precio_costo"].unique()

array([489.9])

In [290]:
df.loc["2020",]

,matricula_funcionario,nombre_funcionario,cargo,codigo_tienda,nombre_tienda,codigo_producto,descripcion_producto,categoria,precio_costo,valor_unitario,cantidad,dt_entrega,valor_final
dt_venta,,,,,,,,,,,,,
2020-08-25 10:04:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10013,Fogón,Electrodomésticos,210.9,404.1,3,2020-08-31 16:39:00,1212.3
2020-06-10 15:34:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10011,Licuadora,Cocina,59.9,142.4,2,2020-06-11 13:59:00,284.8
2020-11-01 11:48:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10018,Colchón,Casa,109.9,249.9,2,2020-11-01 12:31:00,499.8
2020-06-23 10:47:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10018,Colchón,Casa,109.9,249.9,2,2020-06-23 15:56:00,499.8
2020-04-01 16:50:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10010,Freidora eléctrica,Cocina,187.9,340.0,2,2020-04-02 15:20:00,680.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2020-11-02 09:04:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10017,Cama Box,Casa,1189.9,1899.5,2,2020-11-04 12:12:00,3799.0
2020-04-08 16:03:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10016,Guarda Ropa,Casa,489.9,999.0,4,2020-04-13 12:45:00,3996.0
2020-01-29 13:19:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10020,Impresora,Electrónicos,329.9,674.1,2,2020-01-30 12:07:00,1348.2


In [293]:
df.loc[(df["codigo_producto"]== "10015") | (df["codigo_producto"]== "10003"), "codigo_producto"::]

,codigo_producto,descripcion_producto,categoria,precio_costo,valor_unitario,cantidad,dt_entrega,valor_final
dt_venta,,,,,,,,
2020-07-13 15:57:00,10003,Televisor,Electrodomésticos,1400.9,2279.99,2,2020-07-17 15:48:00,4559.98
2019-12-22 13:18:00,10003,Televisor,Electrodomésticos,1400.9,2279.99,2,2019-12-26 16:18:00,4559.98
2020-02-16 16:12:00,10003,Televisor,Electrodomésticos,1400.9,2279.99,2,2020-02-19 11:21:00,4559.98
2019-03-30 11:14:00,10015,Microondas,Electrodomésticos,189.0,364.50,3,2019-03-31 10:21:00,1093.50
2019-04-16 17:44:00,10003,Televisor,Electrodomésticos,1400.9,2279.99,2,2019-04-22 12:04:00,4559.98
...,...,...,...,...,...,...,...,...
2020-08-28 10:08:00,10015,Microondas,Electrodomésticos,189.0,364.50,2,2020-08-28 14:22:00,729.00
2020-03-11 11:08:00,10015,Microondas,Electrodomésticos,189.0,364.50,2,2020-03-11 15:14:00,729.00
2019-06-24 16:01:00,10003,Televisor,Electrodomésticos,1400.9,2279.99,2,2019-06-28 09:22:00,4559.98


## <span style="color: plum">Concepto de Slicing (Rebanado) en Pandas 🍞</span>

El **Slicing** es una técnica que nos permite extraer una "rebanada" o un rango específico de datos de nuestro DataFrame. En lugar de seleccionar elementos uno por uno, definimos un **punto de inicio** y un **punto de fin**.

### 🔍 Anatomía de un Slice
La sintaxis básica utiliza los dos puntos `:` para separar los límites:

> **`rango = [ inicio : fin : paso ]`**

* **Inicio:** Desde dónde empezamos (incluido).
* **Fin:** Hasta dónde llegamos (en etiquetas de Pandas, se incluye el final).
* **Paso:** De cuánto en cuánto saltamos (por defecto es 1).



---

### 🛠️ Análisis de mi código con `.loc`

En la línea que ejecutamos:
`df.loc[(df["codigo_producto"] == "10015") | (df["codigo_producto"] == "10003"), "codigo_producto":]`

#### 1. Filtro de Filas (Antes de la coma)
Usamos el operador OR (`|`) para traer solo los productos con códigos **10015** o **10003**.

#### 2. Slicing de Columnas (Después de la coma)
Al escribir `"codigo_producto":`, aplicamos un slice que dice:
* **Inicio:** Columna `codigo_producto`.
* **Fin:** (Vacío), lo que significa "hasta el final de la tabla".

**Resultado:** El DataFrame nos muestra todas las características de esos productos, pero **oculta** las columnas que estén a la izquierda del código (como IDs internos o fechas que no necesitamos ver en ese momento).

---

### 💡 Atajos rápidos de Slicing

| Código | Significado |
| :--- | :--- |
| `[:]` | Selecciona **todo** el eje (filas o columnas). |
| `[:"precio"]` | Desde el **principio** hasta la columna "precio". |
| `"cantidad":` | Desde "cantidad" hasta el **final**. |
| `::2` | Selecciona uno de cada dos (salta de **2 en 2**). |

## <span style="color: plum">Aplicando Queries ⚡</span>

El método `.query()` es una herramienta de Pandas que nos permite consultar y filtrar columnas utilizando **expresiones booleanas** de una manera mucho más legible, similar al lenguaje natural o a SQL.

### 📝 Sintaxis Básica
A diferencia de `.loc[]`, en una query pasamos toda la condición como una cadena de texto (string).

```python
# Ejemplo: Filtrar productos con valor final mayor a 1000
df.query("valor_final > 1000")

In [294]:
df["categoria"].unique()

<StringArray>
['Electrodomésticos', 'Cocina', 'Electrónicos', 'Casa']
Length: 4, dtype: str

### 🔑 Regla de las Comillas en `.query()`

Para filtrar por texto (strings) dentro de una query, debemos usar **comillas anidadas**:

1. **Comillas Externas:** Delimitan toda la consulta de Pandas.
2. **Comillas Internas:** Delimitan el valor específico que estamos buscando dentro de la columna.

> **Regla de oro:** Si usas comillas dobles `"` para la consulta, usa comillas sencillas `'` para el texto interno (o viceversa). 

* **Correcto:** `df.query("color == 'Rojo'")`
* **Incorrecto:** `df.query("color == Rojo")` -> (Pandas buscará una columna llamada Rojo).

In [297]:
df.query("categoria == 'Electrónicos'")

,matricula_funcionario,nombre_funcionario,cargo,codigo_tienda,nombre_tienda,codigo_producto,descripcion_producto,categoria,precio_costo,valor_unitario,cantidad,dt_entrega,valor_final
dt_venta,,,,,,,,,,,,,
2019-11-25 16:28:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10005,Notebook,Electrónicos,2299.9,3199.00,2,2019-11-26 13:42:00,6398.00
2020-12-10 11:41:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10002,Tablet,Electrónicos,799.0,1424.91,2,2020-12-10 15:53:00,2849.82
2019-12-14 13:44:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10005,Notebook,Electrónicos,2299.9,3199.00,2,2019-12-15 09:06:00,6398.00
2020-05-31 17:46:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10005,Notebook,Electrónicos,2299.9,3199.00,2,2020-06-01 15:56:00,6398.00
2020-02-21 14:19:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10001,Smartphone,Electrónicos,899.9,1676.90,3,2020-02-22 10:54:00,5030.70
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2020-02-27 13:27:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10020,Impresora,Electrónicos,329.9,674.10,2,2020-02-27 15:05:00,1348.20
2019-02-27 15:01:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10005,Notebook,Electrónicos,2299.9,3199.00,4,2019-02-28 14:23:00,12796.00
2019-03-06 15:06:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10020,Impresora,Electrónicos,329.9,674.10,4,2019-03-07 11:39:00,2696.40


In [301]:
df.query("categoria == 'Electrónicos' and matricula_funcionario == '123456999'")

,matricula_funcionario,nombre_funcionario,cargo,codigo_tienda,nombre_tienda,codigo_producto,descripcion_producto,categoria,precio_costo,valor_unitario,cantidad,dt_entrega,valor_final
dt_venta,,,,,,,,,,,,,
2019-11-25 16:28:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10005,Notebook,Electrónicos,2299.9,3199.00,2,2019-11-26 13:42:00,6398.00
2020-12-10 11:41:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10002,Tablet,Electrónicos,799.0,1424.91,2,2020-12-10 15:53:00,2849.82
2019-12-14 13:44:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10005,Notebook,Electrónicos,2299.9,3199.00,2,2019-12-15 09:06:00,6398.00
2020-05-31 17:46:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10005,Notebook,Electrónicos,2299.9,3199.00,2,2020-06-01 15:56:00,6398.00
2020-02-21 14:19:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10001,Smartphone,Electrónicos,899.9,1676.90,3,2020-02-22 10:54:00,5030.70
2019-11-05 17:00:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10005,Notebook,Electrónicos,2299.9,3199.00,2,2019-11-06 13:07:00,6398.00
2019-06-10 09:35:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10001,Smartphone,Electrónicos,899.9,1676.90,2,2019-06-10 10:50:00,3353.80
2019-10-21 12:53:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10020,Impresora,Electrónicos,329.9,674.10,2,2019-10-21 14:21:00,1348.20
2020-03-26 09:49:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10001,Smartphone,Electrónicos,899.9,1676.90,2,2020-03-26 10:41:00,3353.80


## <span style="color: plum">¿Query o Loc? 🤔</span>

| Situación | Ganador | Razón |
| :--- | :--- | :--- |
| **Lectura rápida** | `.query()` | Menos código, más parecido al lenguaje natural. |
| **Recortar columnas** | `.loc[]` | Permite usar el `:` después de la coma. |
| **Cambiar datos** | `.loc[]` | Es el único que permite asignar nuevos valores de forma segura. |
| **Uso de variables** | `.query()` | Muy limpio usando el prefijo `@`. |

# <span style = "Color: Plum"> Agrupamiento de Datos </span>

## <span style = "Color: Plum">groupby </span>

In [306]:
df.groupby("nombre_tienda")["valor_final"].sum().to_frame()

,valor_final
nombre_tienda,
Filial ES,1712534.10
Filial MG,1569556.83
Filial SP,1618746.91
Tienda Matriz,1804008.77


In [310]:
df.groupby("nombre_tienda")[["cantidad", "valor_final"]].sum().round(2)

,cantidad,valor_final
nombre_tienda,,
Filial ES,1792,1712534.10
Filial MG,1824,1569556.83
Filial SP,1799,1618746.91
Tienda Matriz,1945,1804008.77


In [311]:
df.groupby("nombre_tienda").sum(numeric_only=True)

,precio_costo,valor_unitario,cantidad,valor_final
nombre_tienda,,,,
Filial ES,405469.9,683453.91,1792,1712534.10
Filial MG,384307.3,659267.26,1824,1569556.83
Filial SP,394558.2,677568.15,1799,1618746.91
Tienda Matriz,452201.5,760534.74,1945,1804008.77


In [312]:
df.groupby("nombre_funcionario")["cantidad"].sum().to_frame()

,cantidad
nombre_funcionario,
Felipe Mariño,655
Jorge Mateus,597
José María,681
Marcos Coelho,555
Maria del Carmen,667
Mario Enrique,553
Marta Cury,613
Pamela Lima,599
Reinaldo Salcedo,608


In [313]:
df.groupby("nombre_funcionario")[["cantidad"]].sum()

,cantidad
nombre_funcionario,
Felipe Mariño,655
Jorge Mateus,597
José María,681
Marcos Coelho,555
Maria del Carmen,667
Mario Enrique,553
Marta Cury,613
Pamela Lima,599
Reinaldo Salcedo,608


## <span style="color: plum">¿Cuándo usar `()` o `[]`? 🧩</span>

| Símbolo | Nombre | Cuándo usarlo | Ejemplo |
| :--- | :--- | :--- | :--- |
| **`( )`** | **Paréntesis** | Para ejecutar **acciones** (métodos). Van después de verbos. | `.sum()`, `.groupby()`, `.head()` |
| **`[ ]`** | **Corchetes** | Para **seleccionar** columnas o filtrar datos. | `df["columna"]`, `df[df["A"] > 1]` |
| **`[[ ]]`**| **Doble Corchete**| Para seleccionar columnas y mantener el formato de **tabla (DataFrame)**. | `df[["precio", "cantidad"]]` |

> **Regla rápida:** Si le pides a Pandas que **HAGA** algo, usa `()`. Si le pides que **BUSQUE** algo, usa `[]`.

## <span style = "Color: Plum">Agrupando  datos con agg() </span>

In [316]:
df.groupby("descripcion_producto")["valor_final"].agg(["sum","mean","min","max"]) 

,sum,mean,min,max
descripcion_producto,,,,
Aspiradora de piso,83935.80,485.178035,189.90,759.60
Batidora,33085.50,243.275735,95.90,383.60
Cafetera,90883.10,668.258088,284.90,1139.60
Cama Box,685719.50,4696.708904,1899.50,7598.00
Colchón,92463.00,616.420000,249.90,999.60
Fogón,160831.80,986.698160,404.10,1616.40
Freidora eléctrica,137700.00,900.000000,340.00,1360.00
Guarda Ropa,355644.00,2386.872483,999.00,3996.00
Heladera,456599.10,3215.486620,1346.90,5387.60


In [318]:
df.groupby("descripcion_producto").agg(
    {"precio_costo": "sum",
     "valor_final": "mean",}
)


,precio_costo,valor_final
descripcion_producto,,
Aspiradora de piso,15379.7,485.178035
Batidora,6786.4,243.275735
Cafetera,21746.4,668.258088
Cama Box,173725.4,4696.708904
Colchón,16485.0,616.420000
Fogón,34376.7,986.698160
Freidora eléctrica,28748.7,900.000000
Guarda Ropa,72995.1,2386.872483
Heladera,99385.8,3215.486620


## <span style="color: plum">El Poder de `.agg()` (Agregaciones) 📊</span>

El método `.agg()` nos permite realizar múltiples cálculos matemáticos al mismo tiempo después de un `groupby`.

### 1. Agregación por Diccionario `{ }`
Se usa para aplicar **funciones específicas a columnas específicas**.
* **Estructura:** `{"columna": "operacion"}`
* **Ejemplo:** `df.agg({"costo": "sum", "stock": "mean"})`

### 2. Agregación por Lista `[ ]`
Se usa para aplicar **varias funciones a una sola columna** seleccionada.
* **Estructura:** `["columna"].agg(["op1", "op2"])`
* **Ejemplo:** `df["precio"].agg(["min", "max", "std"])`

---

### 💡 Tip Profesional: Renombrar Columnas
Si quieres que el resultado no diga "sum" o "mean" sino algo más claro, puedes hacer esto:
`df.groupby("vendedor").agg(Ventas_Totales=("cantidad", "sum"))`

## <span style = "Color: Plum"> Agrupando datos por Index </span>

In [319]:
df.head()

,matricula_funcionario,nombre_funcionario,cargo,codigo_tienda,nombre_tienda,codigo_producto,descripcion_producto,categoria,precio_costo,valor_unitario,cantidad,dt_entrega,valor_final
dt_venta,,,,,,,,,,,,,
2020-08-25 10:04:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10013,Fogón,Electrodomésticos,210.9,404.1,3,2020-08-31 16:39:00,1212.3
2020-06-10 15:34:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10011,Licuadora,Cocina,59.9,142.4,2,2020-06-11 13:59:00,284.8
2019-03-21 15:58:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10013,Fogón,Electrodomésticos,210.9,404.1,2,2019-03-28 09:44:00,808.2
2019-09-16 14:46:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10004,Máquina de lavar ropa,Electrodomésticos,1099.0,1709.1,2,2019-09-20 09:22:00,3418.2
2019-11-25 16:28:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10005,Notebook,Electrónicos,2299.9,3199.0,2,2019-11-26 13:42:00,6398.0


In [321]:
df.loc["2020"]

,matricula_funcionario,nombre_funcionario,cargo,codigo_tienda,nombre_tienda,codigo_producto,descripcion_producto,categoria,precio_costo,valor_unitario,cantidad,dt_entrega,valor_final
dt_venta,,,,,,,,,,,,,
2020-08-25 10:04:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10013,Fogón,Electrodomésticos,210.9,404.1,3,2020-08-31 16:39:00,1212.3
2020-06-10 15:34:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10011,Licuadora,Cocina,59.9,142.4,2,2020-06-11 13:59:00,284.8
2020-11-01 11:48:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10018,Colchón,Casa,109.9,249.9,2,2020-11-01 12:31:00,499.8
2020-06-23 10:47:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10018,Colchón,Casa,109.9,249.9,2,2020-06-23 15:56:00,499.8
2020-04-01 16:50:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10010,Freidora eléctrica,Cocina,187.9,340.0,2,2020-04-02 15:20:00,680.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2020-11-02 09:04:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10017,Cama Box,Casa,1189.9,1899.5,2,2020-11-04 12:12:00,3799.0
2020-04-08 16:03:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10016,Guarda Ropa,Casa,489.9,999.0,4,2020-04-13 12:45:00,3996.0
2020-01-29 13:19:00,123457113,Victor Hugo,Vendedor,CL004,Filial ES,10020,Impresora,Electrónicos,329.9,674.1,2,2020-01-30 12:07:00,1348.2


In [ ]:
c

,matricula_funcionario,nombre_funcionario,cargo,codigo_tienda,nombre_tienda,codigo_producto,descripcion_producto,categoria,precio_costo,valor_unitario,cantidad,dt_entrega,valor_final
dt_venta,,,,,,,,,,,,,
2020-08-25 10:04:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10013,Fogón,Electrodomésticos,210.9,404.1,3,2020-08-31 16:39:00,1212.3
2020-08-25 17:30:00,123456111,José María,Vendedor,CL001,Tienda Matriz,10004,Máquina de lavar ropa,Electrodomésticos,1099.0,1709.1,2,2020-08-30 11:25:00,3418.2
2020-08-25 14:56:00,123456222,Maria del Carmen,Vendedor,CL001,Tienda Matriz,10011,Licuadora,Cocina,59.9,142.4,4,2020-08-26 10:55:00,569.6
2020-08-25 15:32:00,123456222,Maria del Carmen,Vendedor,CL001,Tienda Matriz,10019,Sofá,Casa,789.9,1399.0,2,2020-08-30 10:12:00,2798.0
2020-08-25 09:11:00,123457111,Pamela Lima,Vendedor,CL004,Filial ES,10019,Sofá,Casa,789.9,1399.0,1,2020-08-28 10:09:00,1399.0
2020-08-25 14:11:00,123456444,Ricardo Lemos,Vendedor,CL002,Filial MG,10008,Cafetera,Cocina,159.9,284.9,3,2020-08-26 13:20:00,854.7


In [ ]:
df.groupby(df.index.year).agg({"valor_final": "sum" })

,valor_final
dt_venta,
2019,3505589.26
2020,3199257.35


In [327]:
df.groupby(df.index.month).agg({"valor_final": "sum" })


,valor_final
dt_venta,
1,536203.08
2,509207.47
3,652415.81
4,491416.26
5,663227.65
6,581055.92
7,459160.47
8,568245.58
9,584067.56


In [328]:
df.groupby(df.index.day).agg({"valor_final": "sum" })


,valor_final
dt_venta,
1,199198.01
2,231860.54
3,238754.65
4,238275.24
5,246657.10
6,171686.01
7,207195.30
8,200640.55
9,256453.99


In [329]:
df.groupby(df.index.hour).agg({"valor_final": "sum" })


,valor_final
dt_venta,
9,692125.21
10,758414.84
11,643785.70
12,720586.27
13,796728.15
14,863890.57
15,767922.80
16,715064.58
17,746328.49


In [332]:
# Agrupamos por hora, sumamos y ordenamos de mayor a menor
df_ventas_hora = df.groupby(df.index.hour).agg({"valor_final": "sum"}).sort_values(by="valor_final", ascending=False)
df_ventas_hora

,valor_final
dt_venta,
14,863890.57
13,796728.15
15,767922.80
10,758414.84
17,746328.49
12,720586.27
16,715064.58
9,692125.21
11,643785.70


## <span style = "Color: Plum"> Agrupando datos por más de una columna </span>

In [ ]:
df.groupby(["nombre_tienda", "nombre_funcionario","descripcion_producto"]).agg({"valor_final": ["sum","mean"] })

valor_final  \
                                                                 sum   
nombre_tienda nombre_funcionario descripcion_producto                  
Filial ES     Mario Enrique      Aspiradora de piso          9495.00   
                                 Batidora                    1150.80   
                                 Cafetera                    3133.90   
                                 Cama Box                   51286.50   
                                 Colchón                     5247.90   
...                                                              ...   
Tienda Matriz Maria del Carmen   Procesador de alimentos     7785.90   
                                 Smartphone                 57014.60   
                                 Sofá                       32177.00   
                                 Tablet                     58421.31   
                                 Televisor                  59279.74   

                                                                       
                                                                 mean  
nombre_tienda nombre_funcionario descripcion_producto                  
Filial ES     Mario Enrique      Aspiradora de piso        527.500000  
                                 Batidora                  230.160000  
                                 Cafetera                  626.780000  
                                 Cama Box                 4662.409091  
                                 Colchón                   524.790000  
...                                                               ...  
Tienda Matriz Maria del Carmen   Procesador de alimentos   486.618750  
                                 Smartphone               4751.216667  
                                 Sofá                     3575.222222  
                                 Tablet                   3245.628333  
                                 Televisor                4559.980000  

[240 rows x 2 columns]

In [350]:
# Prueba agregando ./ al inicio
df.groupby(["nombre_tienda", "nombre_funcionario","descripcion_producto"]).agg({"valor_final": ["sum","mean"] }).to_excel("informe.xlsx")

In [351]:
# QUIERO SABER CUAL ES EL PRODUCTO QUE MAS VENDE EN CADA TIENDA Y CUAL ES EL FUNCIONARIO QUE MAS VENDE CADA PRODUCTO
df.groupby(["nombre_tienda", "nombre_funcionario","descripcion_producto"]).agg({"valor_final": ["sum","mean"] }).to_excel("informe.xlsx")

## <span style = "Color: Plum"> Agrupando datos de tabla dinamicas Pivot Table </span>

In [352]:
df.head()

,matricula_funcionario,nombre_funcionario,cargo,codigo_tienda,nombre_tienda,codigo_producto,descripcion_producto,categoria,precio_costo,valor_unitario,cantidad,dt_entrega,valor_final
dt_venta,,,,,,,,,,,,,
2020-08-25 10:04:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10013,Fogón,Electrodomésticos,210.9,404.1,3,2020-08-31 16:39:00,1212.3
2020-06-10 15:34:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10011,Licuadora,Cocina,59.9,142.4,2,2020-06-11 13:59:00,284.8
2019-03-21 15:58:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10013,Fogón,Electrodomésticos,210.9,404.1,2,2019-03-28 09:44:00,808.2
2019-09-16 14:46:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10004,Máquina de lavar ropa,Electrodomésticos,1099.0,1709.1,2,2019-09-20 09:22:00,3418.2
2019-11-25 16:28:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10005,Notebook,Electrónicos,2299.9,3199.0,2,2019-11-26 13:42:00,6398.0


In [356]:
df.pivot_table(
    index= "descripcion_producto",
    values = ["precio_costo","valor_final"],
    aggfunc=["sum","mean"],
    columns = "nombre_tienda", 
    margins = True,

    )

sum                                    \
                        precio_costo                                     
nombre_tienda              Filial ES Filial MG Filial SP Tienda Matriz   
descripcion_producto                                                     
Aspiradora de piso            4356.1    3822.7    4178.3        3022.6   
Batidora                      1546.9    2095.8    1596.8        1546.9   
Cafetera                      3837.6    4477.2    5916.3        7515.3   
Cama Box                     42836.4   44026.3   38076.8       48785.9   
Colchón                       3626.7    3956.4    4725.7        4176.2   
Fogón                         7592.4   10123.2    8436.0        8225.1   
Freidora eléctrica            6576.5    6576.5    7703.9        7891.8   
Guarda Ropa                  16166.7   20085.9   18616.2       18126.3   
Heladera                     26596.2   27996.0   20297.1       24496.5   
Impresora                    10886.7   12206.3   13855.8       13855.8   
Licuadora                     2276.2    2096.5    2396.0        2575.7   
Microondas                    7560.0    7371.0    8127.0        7560.0   
Máquina de lavar ropa        34069.0   48356.0   45059.0       47257.0   
Notebook                     89696.1   57497.5   52897.7      101195.6   
Plancha                        948.6    1116.0     892.8        1255.5   
Procesador de alimentos       3630.0    5280.0    2970.0        3850.0   
Smartphone                   34196.2   32396.4   35996.0       35996.0   
Sofá                         26066.7   28436.4   30016.2       24486.9   
Tablet                       25568.0   27166.0   31161.0       39950.0   
Televisor                    57436.9   39225.2   61639.6       50432.4   
All                         405469.9  384307.3  394558.2      452201.5   

                                                                        \
                                   valor_final                           
nombre_tienda                  All   Filial ES   Filial MG   Filial SP   
descripcion_producto                                                     
Aspiradora de piso         15379.7    25256.70    20509.20    20889.00   
Batidora                    6786.4     7192.50    10357.20     6808.90   
Cafetera                   21746.4    14814.80    20512.80    25071.20   
Cama Box                  173725.4   186151.00   157658.50   151960.00   
Colchón                    16485.0    18742.50    21741.30    27239.10   
Fogón                      34376.7    33944.40    46875.60    40814.10   
Freidora eléctrica         28748.7    33320.00    33320.00    36380.00   
Guarda Ropa                72995.1    82917.00   100899.00    84915.00   
Heladera                   99385.8   119874.10   130649.30   101017.50   
Impresora                  50804.6    56624.40    51905.70    68758.20   
Licuadora                   9344.4    14952.00    12388.80    12104.00   
Microondas                 30618.0    36814.50    37543.50    35721.00   
Máquina de lavar ropa     174741.0   136728.00   162364.50   182873.70   
Notebook                  301286.9   291109.00   188741.00   166348.00   
Plancha                     4212.9     5142.80     5254.60     4695.60   
Procesador de alimentos    15730.0    16521.30    20509.20    14242.50   
Smartphone                138584.6   174397.60   144213.40   149244.10   
Sofá                      109006.2   104925.00   123112.00   120314.00   
Tablet                    123845.0   118267.53   128241.90   125392.08   
Televisor                 208734.1   234838.97   152759.33   243958.93   
All                      1636536.9  1712534.10  1569556.83  1618746.91   

                                                          mean               \
                                                  precio_costo                
nombre_tienda           Tienda Matriz         All    Filial ES    Filial MG   
descripcion_producto                                                          
Aspiradora de p

## <span style = "Color: Plum"> Iterando sobre un DataFrame </span>



## <span style = "Color: Plum"> iterrows() </span>

In [357]:
df.head()

,matricula_funcionario,nombre_funcionario,cargo,codigo_tienda,nombre_tienda,codigo_producto,descripcion_producto,categoria,precio_costo,valor_unitario,cantidad,dt_entrega,valor_final
dt_venta,,,,,,,,,,,,,
2020-08-25 10:04:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10013,Fogón,Electrodomésticos,210.9,404.1,3,2020-08-31 16:39:00,1212.3
2020-06-10 15:34:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10011,Licuadora,Cocina,59.9,142.4,2,2020-06-11 13:59:00,284.8
2019-03-21 15:58:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10013,Fogón,Electrodomésticos,210.9,404.1,2,2019-03-28 09:44:00,808.2
2019-09-16 14:46:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10004,Máquina de lavar ropa,Electrodomésticos,1099.0,1709.1,2,2019-09-20 09:22:00,3418.2
2019-11-25 16:28:00,123456999,Felipe Mariño,Vendedor,CL003,Filial SP,10005,Notebook,Electrónicos,2299.9,3199.0,2,2019-11-26 13:42:00,6398.0


In [ ]:
df.iterrows()
#sirve para iterar sobre las filas de un DataFrame, devolviendo el índice de cada fila y una Serie con los datos de esa fila. Es útil para realizar operaciones fila por fila, aunque no es la forma más eficiente de trabajar con grandes conjuntos de datos en pandas.

<generator object DataFrame.iterrows at 0x000002C783F1B9A0>

In [362]:
for index, row in df.iterrows():
    print(index)
    print(row)
    break

2020-08-25 10:04:00
matricula_funcionario              123456999
nombre_funcionario             Felipe Mariño
cargo                               Vendedor
codigo_tienda                          CL003
nombre_tienda                      Filial SP
codigo_producto                        10013
descripcion_producto                   Fogón
categoria                  Electrodomésticos
precio_costo                           210.9
valor_unitario                         404.1
cantidad                                   3
dt_entrega               2020-08-31 16:39:00
valor_final                           1212.3
Name: 2020-08-25 10:04:00, dtype: object


In [369]:
for _, row in df.iterrows():
    print(row.descripcion_producto)
    print(row.Valor_final)
    break

Fogón


AttributeError: 'Series' object has no attribute 'Valor_final'


## <span style = "Color: Plum"> itertuples() </span>

In [372]:
for row in df.itertuples():
    print(row)
    print(row.index)
    print(row.descripcion_producto)
    break

Pandas(Index=Timestamp('2020-08-25 10:04:00'), matricula_funcionario='123456999', nombre_funcionario='Felipe Mariño', cargo='Vendedor', codigo_tienda='CL003', nombre_tienda='Filial SP', codigo_producto='10013', descripcion_producto='Fogón', categoria='Electrodomésticos', precio_costo=210.9, valor_unitario=404.1, cantidad=3, dt_entrega=Timestamp('2020-08-31 16:39:00'), valor_final=1212.3000000000002)
<built-in method index of Pandas object at 0x000002C784AD0180>
Fogón


In [373]:
for row in df.itertuples():
    print("Categoría de producto: ", row.categoria)
    print("Descripción del producto: ", row.descripcion_producto)
    print("Valor final venta de producto: ", row.valor_final)
    print("\n")

Categoría de producto:  Electrodomésticos
Descripción del producto:  Fogón
Valor final venta de producto:  1212.3000000000002


Categoría de producto:  Cocina
Descripción del producto:  Licuadora
Valor final venta de producto:  284.8


Categoría de producto:  Electrodomésticos
Descripción del producto:  Fogón
Valor final venta de producto:  808.2


Categoría de producto:  Electrodomésticos
Descripción del producto:  Máquina de lavar ropa
Valor final venta de producto:  3418.2


Categoría de producto:  Electrónicos
Descripción del producto:  Notebook
Valor final venta de producto:  6398.0


Categoría de producto:  Casa
Descripción del producto:  Colchón
Valor final venta de producto:  499.8


Categoría de producto:  Cocina
Descripción del producto:  Freidora eléctrica
Valor final venta de producto:  680.0


Categoría de producto:  Casa
Descripción del producto:  Colchón
Valor final venta de producto:  499.8


Categoría de producto:  Casa
Descripción del producto:  Colchón
Valor final ve